Semi specific digestion

Jeremi-Maciejewski opened the pull request:
## Description

These changes implement support of performing semi-specific digestion to ProteaseDigestion, as requested in #8098 .
Also exposed via pyOpenMS.

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for semi-specific digestion in protease digestion, allowing generation of peptides with only one end at a cleavage site when specificity is set accordingly.

* **Documentation**
  * Updated user documentation and code comments to describe semi-specific digestion functionality and provide usage examples.

* **Tests**
  * Introduced new unit tests to verify semi-specific digestion and related behaviors.
  * Removed outdated tests for protease digestion.
  * Added test coverage for error handling and edge cases in semi-specific digestion.

* **Chores**
  * Added a new contributor to the authors list.
  * Updated digestion specificity in decoy peptide generation to enforce full enzymatic specificity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Add method handling semi-specific digestion to EnzymaticDigestion and use it within its child Protease Digestion; expand ProteaseDigestion_test; expand ProteaseDigestion python binding test and move it from test000.py to new test_ProteaseDigestion.py

Commit:
Updated/added authors section to some files

Commit:
Merge branch 'develop' of github.com:OpenMS/OpenMS into Semi-specific_digestion

Commit:
Update CHANGELOG.

cbielow commented on src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h:
```diff
@@ -179,6 +179,21 @@ namespace OpenMS
+    /**
+      @brief Generates semi-specific digestion products
+
+      Function handling semi-specific digestion (specificity == 1).
```
```suggestion
      Function handling semi-specific digestion (specificity_ == Specificity::SPEC_SEMI ).
```

Jeremi-Maciejewski replied:
Corrected.

cbielow commented on src/openms/include/OpenMS/CHEMISTRY/ProteaseDigestion.h:
```diff
@@ -49,7 +49,9 @@ namespace OpenMS
     /** 
-       @brief Performs the enzymatic digestion of a protein represented as AASequence
+       @brief Performs the enzymatic digestion of a protein represented as AASequence.
+       By default, only fully specific products are generated.
+       If object's Specificity is set to SPEC_SEMI (1), also generates semi-specific products.
```
There is no `default` here (when function parameters are concerned).
-->
```
Digestion takes into account the value of
 - the **missed cleavages** member, see @p setMissedCleavages()
 - the **specificity** member, see @p setSpecificity()

Currently, Specificity::SPEC_SEMI and Specificity::SPEC_FULL are supported. Others raise an exception

@throw Exception::InvalidValue   // <-- todo

```

```

cbielow replied:
please also update the docs of `Size digest(const AASequence& protein, ....` since that is just a wrapper.
(maybe just add a doxgen reference pointing here for details)

cbielow commented on src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h:
```diff
@@ -179,6 +179,21 @@ namespace OpenMS
     std::vector<int> tokenize_(const String& sequence, int start = 0, int end = -1) const;
 
+    /**
+      @brief Generates semi-specific digestion products
```
will it skip fully specific variants?

will it take MC's into account?

Please document

Jeremi-Maciejewski replied:
Done.

cbielow commented on src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:
```diff
@@ -115,6 +115,65 @@ namespace OpenMS
+      if (min_length <= std::abs((int)y - (int)x) &&
+          max_length >= std::abs((int)y - (int)x))
+      {
+        output.emplace_back(x, y);
```
since x < y by this definition, do you need std::abs above?

Jeremi-Maciejewski replied:
Not really, I removed it.

cbielow commented on src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:
```diff
@@ -115,6 +115,65 @@ namespace OpenMS
+
+    auto fwd = cleavage_positions.cbegin();
+    auto bwd = cleavage_positions.crbegin(); // Reverse iterator!
+    ++fwd; ++bwd;
```
dangerous!
check `cleavage_positions.empty()` and return as a first action in the function.

Jeremi-Maciejewski replied:
Implemented (throws `Exception::InvalidValue` on too few cleavage positions - should actually be at least 2)

cbielow commented on src/openms/source/CHEMISTRY/ProteaseDigestion.cpp:
```diff
@@ -139,6 +139,13 @@ namespace OpenMS
     }
+
+    // semi-specific variants
+    if (specificity_ == SPEC_SEMI)
```
test for specificity_  and throw exception unless its SEMI or FULL

Jeremi-Maciejewski replied:
Done.

cbielow commented on src/pyOpenMS/tests/unittests/test_ProteaseDigestion.py:
```diff
@@ -0,0 +1,56 @@
+    ff.setSpecificity(1) # Semi-specific
+    output = []
+    ff.digest(seq, output)
+    # MHAR, LVP, HAR, LV, AR, L, R, MHA, VP, MH, P, M
+    assert len(output) == 12
```
test the new exception for e.g., Specificity UNKNOWN

Jeremi-Maciejewski replied:
Done. (although with SIZE_OF_SPECIFICITY)

cbielow commented on src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp:
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
+
+    // Lambda to allow searching 'out' with std::find_if()
+    std::string ref = "";
+    auto compare_aa = [&ref](AASequence seq) -> bool
```
```suggestion
    auto compare_aa = [&ref](const AASequence& seq) -> bool
```

Jeremi-Maciejewski replied:
Corrected.

cbielow commented on src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp:
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
+    TEST_EQUAL(out[1].toString(), "MHP");
+    // Are there semi-specific variants?
+    ref = "AGL";
+    TEST_EQUAL(std::find_if(out.begin(),out.end(),compare_aa) != out.end(), true);
```
you could use TEST_TRUE here

Jeremi-Maciejewski replied:
I used TEST_EQUAL since its documentation says:
`If possible use TEST_EQUAL(a, b) instead of TEST_TRUE(a==b), because the latter makes bug tracing harder.`

(same for TEST_FALSE and TEST_NOT_EQUAL)

cbielow replied:
we should update the docs, sorry for the confusion.
it should say, `if you test for a boolean, use TEST_TRUE / TEST_FALSE, otherwise prefer TEST_EQUAL/TEST_NOT_EQUAL to ease bug tracing.`
E.g. do not do this
`TEST_EQUAL(myStringFunc() == "some_return_value", true)`, because it will fail, but not say what the actual return value was

Jeremi-Maciejewski replied:
I see, that does make more sense. Thanks for explaining, I'll change it to `TEST_TRUE`/`TEST_FALSE` .

Jeremi-Maciejewski replied:
Corrected

cbielow commented on src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp:
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
+    // Make sure cleavage sites are iterated over in right order.
+    // (forward for N-terminus cleavage, backwards for C-terminus cleavage)
+    ref = "GLRMHP";
+    TEST_NOT_EQUAL(std::find_if(out.begin(),out.end(),compare_aa) != out.end(), true);
```
TEST_FALSE? :)

Jeremi-Maciejewski replied:
Corrected.

cbielow commented on src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp:
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
+    TEST_EQUAL(out.size(), 4*3/2);
+
+    pd.setEnzyme("Trypsin/P");
+    pd.digest(AASequence::fromString("AGLRMHP"), out);
```
can you test this with a protein where the MC setting is less than the actual number of more cleavage sites?! Just to make sure this is done correctly.
Also, the other way around: A protein with no cleavage sites and MC=1 should produce valid output.

Jeremi-Maciejewski replied:
First was already implemented, I added test for 2nd case.

cbielow reviewed (changes requested):
nice addition, thanks!

some comments below

Commit:
Applied suggestions from PR review (#8130).
Expanded and reworded documentation, added some input data checks and exceptions to `ProteaseDigestion.digest()` and `EnzymaticDigestion.semiSpecificDigestion_()`, added some tests.

Commit:
Merge branch 'develop' of github.com:OpenMS/OpenMS into Semi-specific_digestion

Commit:
Added missing import to PyOpenMS test;
Fixed invalid (not implemented)  use of SPEC_NONE with `ProteaseDigestion::digest()` by functions `DecoyGenerator::reversePeptides()` and `DecoyGenerator::shufflePeptides()`.
These now use `SPEC_FULL`, which should produce identical results as before - but perhaps this was not the behaviour intended?

Commit:
Merge branch 'develop' of github.com:OpenMS/OpenMS into Semi-specific_digestion

Commit:
Typos; Code readability and correcting unrecommended use of boolean comparisons in tests.

Commit:
`test_ProteaseDigestion.py`: corrected minor error, use Python's operator `not` for negation.

cbielow commented on src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:
```diff
@@ -115,6 +115,73 @@ namespace OpenMS
+    if (cleavage_positions.size() < 2)
+    {
+      String value(cleavage_positions.begin(), cleavage_positions.end());
+      throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION,
+        String("Too few cleavage positions - at least sequence start and end positions are required."), value);
```
document that using Doxygen

Jeremi-Maciejewski replied:
Done.

cbielow commented on src/tests/class_tests/openms/source/EnzymaticDigestion_test.cpp:
```diff
@@ -282,6 +282,33 @@ START_SECTION((Size digestUnmodified(const StringView sequence, std::vector<Stri
+    class TempChild : public EnzymaticDigestion
+    {
+        public:
+            void tmpSetEnzyme(const DigestionEnzyme* enzyme)
+            {
+                this->setEnzyme(enzyme);
+            }
```
do you need this overload?
`setEnyme()` is public.

Jeremi-Maciejewski replied:
Right, I do not. Removed.

cbielow commented on CHANGELOG:
```diff
@@ -29,6 +29,9 @@ Dependencies:
 NucleicAcidSearchEngine:
     - add support for global fixed modifications as a search parameter
 
+ProteaseDigestion:
+    - add support for semi-specific digestion in objects where specificity is set to SPEC_SEMI (1)
```
move this to  Library and mention PR number:
```
Library:
 ...
  - ProteaseDigestion: add support for semi-specific digestion (#8130)
```

Jeremi-Maciejewski replied:
Done.

cbielow reviewed (commented):
nice. almost there :)

Commit:
CHANGELOG wording; check whether input is sorted in semiSpecificDigestion_; missing documentation; additional tests; fix pyOpenMS tests

Commit:
Fix failing pyOpenMS test.
